modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -151,11 +149,20 @@ private void adjustMaxBytesPerGatheringWrite(int attempted, int written, int old
     private static void copyBytes(ByteBuffer[] source, int nioBufferCnt, ByteBuffer destination) {
         for (int i = 0; i < nioBufferCnt && destination.hasRemaining(); i++) {
             ByteBuffer buffer = source[i];
+            assert buffer.hasArray() : "Buffer must have heap array";
             int nBytesToCopy = Math.min(destination.remaining(), buffer.remaining());
-            int initialLimit = buffer.limit();
-            buffer.limit(buffer.position() + nBytesToCopy);
-            destination.put(buffer);
-            buffer.limit(initialLimit);
+            destination.put(buffer.array(), buffer.arrayOffset() + buffer.position(), nBytesToCopy);
+        }
+    }
+
+    private static void setWrittenBytes(ByteBuffer[] source, int bytesWritten) {
+        int i = 0;
```

Comment:
NIT: maybe keep the style we use in `copyBytes` for the loop to keep this a little shorter and easier to digest:

```
        for (int i = 0; bytesWritten > 0 && i < source.length; i++) {
            ByteBuffer buffer = source[i];
            int nBytes = Math.min(buffer.remaining(), bytesWritten);
            buffer.position(buffer.position() + nBytes);
            bytesWritten -= nBytes;
        }
```

Reply:
I adopted this loop style without the `source.length` boundary condition. I don't think that boundary should be necessary as bytes written should not exceed the bytes in the buffers.